rust/agama-server/src/software/web.rs

```diff
@@ -153,6 +230,74 @@ async fn products(State(state): State<SoftwareState<'_>>) -> Result<Json<Vec<Pro
     Ok(Json(products))
 }
 
+/// Information about registration configuration (product, patterns, etc.).
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct RegistrationInfo {
+    /// Registration key. Empty value mean key not used or not registered.
+    key: String,
+    /// Registration email. Empty value mean email not used or not registered.
+    email: String,
+    /// if registration is required, optional or not needed for current product.
+    /// Change only if selected product is changed.
+    requirement: RegistrationRequirement,
+}
+
+/// returns registration info
+///
+/// * `state`: service state.
+#[utoipa::path(get, path = "/software/registration", responses(
+    (status = 200, description = "registration configuration", body = RegistrationInfo),
+    (status = 400, description = "The D-Bus service could not perform the action")
+))]
+async fn get_registration(
+    State(state): State<SoftwareState<'_>>,
+) -> Result<Json<RegistrationInfo>, Error> {
+    let result = RegistrationInfo {
+        key: state.product.registration_code().await?,
+        email: state.product.email().await?,
+        requirement: state.product.registration_requirement().await?,
+    };
+    Ok(Json(result))
+}
+
+/// Software service configuration (product, patterns, etc.).
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct RegistrationParams {
+    /// Registration key.
+    key: String,
+    /// Registration email.
+    email: String,
+}
+
+/// Register product
+///
+/// * `state`: service state.
+#[utoipa::path(post, path = "/software/registration", responses(
+    (status = 200, description = "registration successfull"),
+    (status = 400, description = "The D-Bus service could not perform the action")
+))]
+async fn register(
+    State(state): State<SoftwareState<'_>>,
+    Json(config): Json<RegistrationParams>,
+) -> Result<Json<()>, Error> {
+    state.product.register(&config.key, &config.email).await?;
+    Ok(Json(()))
```

Comment:
Following your own suggestion, I would return `Ok(StatusCode::NO_CONTENT)` (you need to change the return type to `Result<impl IntoResponse, Error>`). We avoid then returning a useless `null`. The same for the `deregister` method below.

Reply:
well, for me reason is that POST has mandatory response body according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST

And that reminds me that what is still missing from old API is more detailed error codes. I am not sure if we can have something like `Ok(Json(scc_error))`?